e2e/tests.go

```diff
@@ -0,0 +1,298 @@
+// Copyright 2023 Sauce Labs Inc. All rights reserved.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/saucelabs/forwarder/e2e/testrunner"
+)
+
+const (
+	proxyService    = "proxy"
+	upstreamService = "upstream-proxy"
+	HttpbinService  = "httpbin"
+)
+
+type testMod func(t *testrunner.Test)
+
+func withName(name string) testMod {
+	return func(t *testrunner.Test) {
+		t.Name = name
+	}
+}
+
+func withCallback(env ...string) testMod {
+	return func(t *testrunner.Test) {
+		t.Callback = func() error {
+			cmd := exec.Command("make", "test")
+			cmd.Env = append(os.Environ(), env...)
+			var stdout, stderr bytes.Buffer
+			cmd.Stdout = &stdout
+			cmd.Stderr = &stderr
+			err := cmd.Run()
+			if err != nil {
+				log.Printf("stdout: %s", stdout.String())
+				log.Printf("stderr: %s", stderr.String())
+			} else {
+				if stderr.Len() > 0 {
+					log.Printf("%s", stderr.String())
+				}
+				s := strings.Split(stdout.String(), "\n")
+				for _, l := range s {
+					if strings.HasPrefix(l, "---") {
+						log.Printf("%s", l)
+					}
+				}
+			}
+			return err
+		}
+	}
+}
+
+func withProtocols(proxy, httpbin string) testMod {
+	return func(t *testrunner.Test) {
+		for _, s := range t.Compose {
+			if s.Name == "proxy" {
+				s.Environment["FORWARDER_PROTOCOL"] = proxy
+			} else if s.Name == "httpbin" {
+				s.Environment["FORWARDER_PROTOCOL"] = httpbin
+			}
+		}
+	}
+}
+
+func withBasicAuth(auth string) testMod {
+	return func(t *testrunner.Test) {
+		for _, s := range t.Compose {
```

Comment:
Looking at how you have to iterate Compose should have a Service(name string) helper.